### PR TITLE
Ppu regs

### PR DESCRIPTION
**Pull request by kingcons**
Most of the core PPU register behavior is here and tested. The missing piece before it's ready for merge is writing to VRAM which I'll code in the morning.

We'll still need to make higher level helpers for use during rendering to construct tiles and sprites, modify scrolling, etc. But that can wait until these branches are reviewed and merged. :)

(Edit note: Oh yeah, and nametable mirroring!)

### CODE REVIEW

**Comment by kingcons** [comment]
Okay, I think this is good for merge after your review.

File: __tests__/ppu_test.re [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    expect(reader(regs)) |> toEqual(true);
+  });
+};
+
+describe("PPU", () => {
+  let ppu = Ppu.build(nestest);
+  let regs = ppu.registers;
+
+  describe("PPUCTRL", () => {
+    ctrl_test_helper(regs, 0, Ppu.x_scroll_offset, 0, 256);
```
Comment (jamesdabbs) [inline __tests__/ppu_test.re]: I think it'd be slightly easier to scan if this helper were defined inside the `decribe("PPUCTRL", { ... })` block. FWIW, you could close over `regs` in that case. (Sim. for `PPUMASK` / `mask_test_helper` below)

Comment (kingcons) [reply __tests__/ppu_test.re]: That's a good note. I'll reshuffle before merging!

File: src/Ppu.re [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+      write_latch: false
+    },
+    oam: Array.make(0x100, 0),
+    name_table: Array.make(0x800, 0),
+    palette_table: Array.make(0x20, 0),
+    pattern_table: rom
+  };
+};
+
+let ctrl_helper = (n, unset, set) => {
```
Comment (jamesdabbs) [inline src/Ppu.re]: Couldn't this be `let ctrl_helper = (n, unset, set, regs) => ...`?

File: src/Ppu.re [hunk 3]
```diff
@@ -0,0 +1,164 @@
+type registers = {
+  mutable control: int,
+  mutable mask: int,
+  mutable status: int,
+  mutable oam_address: int,
```
Comment (jamesdabbs) [inline src/Ppu.re]: Aside - I was thinking about adding something like 

```
type address = int
type byte = int
// others?
```

to `Types.re`. I'm not sure if there's a type-aliasing mechanism that would prevent you from inadvertently assigning one to the other, but I think it'd still be nice to be able to declare your intent a little more precisely.

File: src/Ppu.re [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+let write_vram = (ppu, value) => {
+  let address = ppu.registers.ppu_address;
+  if (address < 0x2000) {
+    Bytes.set(ppu.pattern_table.chr, address, Char.chr(value));
+  } else if (address < 0x3f00) {
+    Array.set(ppu.name_table, address land 0x7ff, value);
+  } else {
+    Array.set(ppu.palette_table, address land 0x1f, value);
+  };
```
Comment (jamesdabbs) [inline src/Ppu.re]: I'd like to take a look at this when we look at memory mapping in general. It feels like we have a `vram` lens, and then a higher-level stateful accessor that reads from `vram` and updates the `ppu_address`.

A big part of why I want to extract all these lenses is I think we can write some higher-level tests that say something like "this lens can read and write at all of these locations".

File: src/Ppu.re (hunk 3 above)
Comment (kingcons) [reply src/Ppu.re]: I definitely hear this. My frustration is that I want types that can constrict input to certain ranges and I feel the lack of such types is visible in all the places that we try to wrap around, i.e. `something land 0x1f`. We could create named types for our different address spaces like:
```
type PaletteIndex
type OAMIndex
type NametableIndex
type PRGIndex
type CHRIndex
etc
```
and then have them wrap during construction but that also feels maybe a bit heavyweight? I dunno. It does seem like something we'll have time to revisit.

File: src/Ppu.re (hunk 2 above)
Comment (kingcons) [reply src/Ppu.re]: Ah, the partial application trick! Yes, thank you. :)

File: src/Ppu.re (hunk 4 above)
Comment (kingcons) [reply src/Ppu.re]: Yeah, I don't have as clear a view on how this would look but I'm definitely interested.  :metal:

**Pull request merged**